block/stateupdate_validator.go

```diff
@@ -0,0 +1,162 @@
+package block
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/dymensionxyz/dymint/da"
+	"github.com/dymensionxyz/dymint/settlement"
+	"github.com/dymensionxyz/dymint/types"
+)
+
+// StateUpdateValidator is a validator for messages gossiped in the p2p network.
+type StateUpdateValidator struct {
+	logger       types.Logger
+	blockManager *Manager
+}
+
+// NewValidator creates a new Validator.
+func NewStateUpdateValidator(logger types.Logger, blockManager *Manager) *StateUpdateValidator {
+	return &StateUpdateValidator{
+		logger:       logger,
+		blockManager: blockManager,
+	}
+}
+
+func (v *StateUpdateValidator) ValidateStateUpdate(batch *settlement.ResultRetrieveBatch) error {
+	v.logger.Debug("validating state update", "start height", batch.StartHeight, "end height", batch.EndHeight)
+
+	err := v.validateDRS(batch.StartHeight, batch.EndHeight, batch.DRSVersion)
+	if err != nil {
+		return err
+	}
+
+	var daBlocks []*types.Block
+	var p2pBlocks []*types.Block
+
+	// load blocks for the batch height, either P2P or DA blocks
+	for height := batch.StartHeight; height <= batch.EndHeight; height++ {
+		source, err := v.blockManager.Store.LoadBlockSource(height)
+		if err != nil {
+			continue
+		}
+		block, err := v.blockManager.Store.LoadBlock(height)
+		if err != nil {
+			continue
+		}
+		if source == types.DA.String() {
+			daBlocks = append(daBlocks, block)
+		} else {
+			p2pBlocks = append(p2pBlocks, block)
+		}
+	}
+
+	if uint64(len(daBlocks)) != batch.NumBlocks {
+		daBlocks = []*types.Block{}
+		var daBatch da.ResultRetrieveBatch
+		for {
+			daBatch = v.blockManager.Retriever.RetrieveBatches(batch.MetaData.DA)
+			if daBatch.Code == da.StatusSuccess {
+				break
+			}
+		}
+		for _, batch := range daBatch.Batches {
+			daBlocks = append(daBlocks, batch.Blocks...)
+		}
```

Comment:
alghough right now da only supports one batch per state update, it could be that in the future there are  multiple da batches (blobs) in the same state update.